### PR TITLE
feat(core): implement generic namespace of exceptions

### LINKED ISSUE #13528: Revisit exception namespacing
Map all PHP exceptions to Elgg namespaces versions https://www.php.net/manual/en/spl.exceptions.php

Maybe add a generic interface to all base exception extensions so all exceptions are known to be from Elgg and can be catched as such. See https://www.alainschlesser.com/structuring-php-exceptions/

### PR DESCRIPTION
**Pull request by jeabakker**
More base PHP exception classes are now present in Elgg and a generic
interface was added to catch all exceptions thrown from the Elgg system.

fixes #13528

### CODE REVIEW

**Comment by jeabakker** [comment]
more follow up in #14074

File: engine/classes/Elgg/Database/Entities.php [hunk 1]
```diff
@@ -189,15 +188,15 @@ public function getDates() {
 	/**
 	 * Execute the query resolving calculation, count and/or batch options
 	 *
-	 * @return array|\ElggData[]|ElggEntity[]|false|int
+	 * @return array|\ElggData[]|\ElggEntity[]|false|int
 	 * @throws \LogicException
```
Comment (jdalsem) [inline engine/classes/Elgg/Database/Entities.php:192]: found one

File: engine/classes/Elgg/Database/SiteSecret.php [hunk 2]
```diff
@@ -50,7 +51,7 @@ public function __construct($key) {
 	 */
```
Comment (jdalsem) [inline engine/classes/Elgg/Database/SiteSecret.php:51]: document throw?

File: engine/classes/Elgg/Filesystem/Directory/Fly.php [hunk 3]
```diff
@@ -85,7 +85,7 @@ public function getContents($path) {
 	 */
 	public function getFile($path) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Filesystem/Directory/Fly.php:86]: throw docs? or is it already part of the inherited docs?

File: engine/classes/Elgg/ImageService.php [hunk 4]
```diff
@@ -9,6 +9,7 @@
 use Imagine\Image\Box;
 use Imagine\Image\ImagineInterface;
 use Imagine\Image\Point;
+use Elgg\Exceptions\LogicException;
```
Comment (jdalsem) [inline engine/classes/Elgg/ImageService.php:12]: position

File: engine/classes/Elgg/PasswordService.php [hunk 5]
```diff
@@ -15,7 +17,7 @@ final class PasswordService {
 	 */
 	public function __construct() {
```
Comment (jdalsem) [inline engine/classes/Elgg/PasswordService.php:18]: throw docs?

File: engine/classes/Elgg/Traits/Database/LegacyQueryOptionsAdapter.php [hunk 6]
```diff
@@ -204,7 +204,7 @@ protected function normalizeTypeSubtypeOptions(array $options = []) {
 				}
 			}
 		} else if (isset($options['subtypes'])) {
-			throw new \Elgg\Exceptions\InvalidArgumentException('If filtering for entity subtypes it is required to provide one or more entity types.');
```
Comment (jdalsem) [inline engine/classes/Elgg/Traits/Database/LegacyQueryOptionsAdapter.php:207]: missing throw docs

File: engine/classes/Elgg/Traits/Seeding.php [hunk 7]
```diff
@@ -134,7 +135,7 @@ public function createUser(array $attributes = [], array $metadata = [], array $
 
 				$user = get_user($guid);
 				if (!$user instanceof \ElggUser) {
-					throw new \Exception("Unable to create new user with attributes: " . print_r($attributes, true));
+					throw new Exception("Unable to create new user with attributes: " . print_r($attributes, true));
```
Comment (jdalsem) [inline engine/classes/Elgg/Traits/Seeding.php:138]: throws docs?

File: engine/classes/Elgg/Upgrade/Loop.php [hunk 8]
```diff
@@ -74,7 +75,7 @@ public function __construct(
 		// Get the class taking care of the actual upgrading
 		$this->batch = $upgrade->getBatch();
 		if (!$this->batch) {
-			throw new \RuntimeException(elgg_echo('admin:upgrades:error:invalid_batch', [
+			throw new RuntimeException(elgg_echo('admin:upgrades:error:invalid_batch', [
```
Comment (jdalsem) [inline engine/classes/Elgg/Upgrade/Loop.php:78]: missing throws docs

File: engine/lib/configuration.php [hunk 9]
```diff
@@ -15,6 +15,7 @@
  */
 
 use Elgg\Project\Paths;
+use Elgg\Exceptions\Exception as ElggException;
```
Comment (jdalsem) [inline engine/lib/configuration.php:18]: order

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Database/SiteSecret.php (hunk 2 above)
Comment (jeabakker) [reply engine/classes/Elgg/Database/SiteSecret.php:51]: yep

File: engine/classes/Elgg/Filesystem/Directory/Fly.php (hunk 3 above)
Comment (jeabakker) [reply engine/classes/Elgg/Filesystem/Directory/Fly.php:86]: added

File: engine/classes/Elgg/Traits/Database/LegacyQueryOptionsAdapter.php (hunk 6 above)
Comment (jeabakker) [reply engine/classes/Elgg/Traits/Database/LegacyQueryOptionsAdapter.php:207]: added

File: engine/classes/Elgg/Database/River.php [hunk 10]
```diff
@@ -100,7 +101,7 @@ public function count() {
 	public function calculate($function, $property, $property_type = 'annotation') {
```
Comment (jdalsem) [inline engine/classes/Elgg/Database/River.php:101]: missing throw

**Review by jdalsem — APPROVED** [review]

**Review by jdalsem — APPROVED** [review]

**Pull request merged**